jax/_src/scipy/signal.py

```diff
@@ -146,3 +148,351 @@ def detrend(data, axis=-1, type='linear', bp=0, overwrite_data=None):
       coef, *_ = linalg.lstsq(A, data[sl])
       data = data.at[sl].add(-jnp.matmul(A, coef, precision=lax.Precision.HIGHEST))
     return jnp.moveaxis(data.reshape(shape), 0, axis)
+
+
+def _triage_segments(window, nperseg, input_length):
+  """
+  Parses window and nperseg arguments for spectrogram and _spectral_helper.
+  This is a helper function, not meant to be called externally.
+  Parameters
+  ----------
+  window : string, tuple, or ndarray
+      If window is specified by a string or tuple and nperseg is not
+      specified, nperseg is set to the default of 256 and returns a window of
+      that length.
+      If instead the window is array_like and nperseg is not specified, then
+      nperseg is set to the length of the window. A ValueError is raised if
+      the user supplies both an array_like window and a value for nperseg but
+      nperseg does not equal the length of the window.
+  nperseg : int
+      Length of each segment
+  input_length: int
+      Length of input signal, i.e. x.shape[-1]. Used to test for errors.
+  Returns
+  -------
+  win : ndarray
+      window. If function was called with string or tuple than this will hold
+      the actual array used as a window.
+  nperseg : int
+      Length of each segment. If window is str or tuple, nperseg is set to
+      256. If window is array_like, nperseg is set to the length of the
+      6
+      window.
+  """
+  # parse window; if array like, then set nperseg = win.shape
+  if isinstance(window, (str, tuple)):
+    # if nperseg not specified
+    if nperseg is None:
+      nperseg = 256  # then change to default
+    if nperseg > input_length:
+      warnings.warn(f'nperseg = {nperseg} is greater than input length '
+                    f' = {input_length}, using nperseg = {nperseg}')
+      nperseg = input_length
+    win = osp_signal.get_window(window, nperseg)
+  else:
+    win = jnp.asarray(window)
+    if len(win.shape) != 1:
+      raise ValueError('window must be 1-D')
+    if input_length < win.shape[-1]:
+      raise ValueError('window is longer than input signal')
+    if nperseg is None:
+      nperseg = win.shape[0]
+    elif nperseg is not None:
+      if nperseg != win.shape[0]:
+        raise ValueError("value specified for nperseg is different"
+                          " from length of window")
+  return win, nperseg
+
+def _fft_helper(x, win, detrend_func, nperseg, noverlap, nfft, sides):
+  """Internal helper function for calculating windowed FFT.
+
+  This funciton is basically a jax re-implementation of
+  https://github.com/scipy/scipy/blob/47bb6febaa10658c72962b9615d5d5aa2513fa3a/scipy/signal/spectral.py#L1874
```

Comment:
If the code is copied & adapted from there, we need to put it in https://github.com/google/jax/tree/main/jax/_src/third_party/scipy

Reply:
I think this function itself is sufficiently different from the original one (Jax-reimplementation is non-trivial). However, the `_triage_segments` above is a literal copy, so it is moved to "third_party".